### DISCUSSION TITLE
Simulist- an R package to simulate linelists

### DISCUSSION THREAD

**Discussion by CarmenTamayo** [epiverse-trace/epiverse-trace.github.io#46] [Project Ideas]
# _Simulist_: tools to generate simulated linelist data

## Description

This R package would provide tools to simulate epidemic's raw case data in the form of line lists, using model outputs, such as compartmental models, branching processes, or network models.

*Potential feature: this package could also include a function to simulate contact tracing data that complements the generated line lists.

R packages such as _outbreak_  or _ringbp_ , offer the possibility to simulate outbreak data (e.g., epicurves) and/or provide the user with a number of already simulated line lists to work with. However, currently available packages do not have the functionality to simulate line list data from scratch.

_Simulist_ would allow users to generate line lists that are tailored to their specific circumstances, by easily being able to modify the number of epidemic cases, the dates of disease onset and case hospitalisation/death, as well as exposure dates. This way, the user would have full control over the simulated raw data and be able to incorporate changes as their models develop.

*If the functionality to generate contact tracing data is included in the package, end-users would be able to generate a contact database that matches the cases included in the simulated line list.

## Target audience

* typical end-users: epidemiologists that want to simulate raw case data from an epidemic, either to test interventions such as contact tracing strategies or develop training materials for students or for academic use, e.g., workshops.

* potential contributors: members of the Epiverse team

* key collaborators: end-users to provide use cases and improve the functionality of the package

## Interoperability

* inputs: a `data.frame` or `tibble` that contains the specified model's output, which must include information about the epidemic's incidence over time.

* outputs: a `data.frame` or `tibble` where rows represent epidemic cases and columns contain information about their dates of symptom onset and potential hospitalisations or deaths.

* imports:
    * _epiparameter_
    * _bpmodels_
    * _epirecipes_
    * ...

* used by:
    * _linelist_
    * _readepi_
    * ...

## Usage

The code below illustrates a typical use of the package, using fictitious code and outputs if needed:

```r
library(simulist)

# Use case 1: linelist simulation
model_output %>%
  add_delays() %>%
  convert_to_dates() %>%
  add_demographic_info(age, sex) %>%
  add_case_info(type, ct value) %>%
  add_case_names()


# Use case 2: contact database
simulated_linelist %>%
  extract_cases() %>%
  add_contacts(contact characteristics specification) %>%
  add_exposure_dates() %>%
  add_contact_names()

```

**Comment by jamesmbaazam** [comment]
Hi Carmen, I believe this is a great idea and I'm actually quite surprised that such a package doesn't already exist. 

I think this package should make it easy to quickly (ease of use) and speedily (fast) synthesise the data. Building on the previous point, I, therefore, think that the piped operations you suggested under use-case 1 should rather run under the hood. An example set of operations under use-case 1 could be
```
case_time_series %>%
    expand_to_linelist() %>%
    summarise() %>%
    communicate()

```

**Comment by jamesmbaazam** [comment]
This potential package could address the [issue](https://github.com/epiverse-trace/bpmodels/issues/9#issue-1459860950) in `bpmodels`.

**Comment by jamesmbaazam** [comment]
There is existing functionality out there that we might want to consider in developing this package:
- [epidist](https://github.com/R4EPI/epidict) via [epidist::gen_data()](https://github.com/R4EPI/epidict/blob/master/R/gen_data.R)
- [EpiNow](https://github.com/epiforecasts/EpiNow/blob/master/R/linelist_tools.R)
- simulacr via [simulacr::simulate_outbreak()](https://github.com/reconhub/simulacr/blob/master/R/simulate_outbreak.R)

Most of these are either in dev but unstable or no longer in dev so, there is a case for developing `simulist` as it could be useful for internal use, teaching, and dev elsewhere.

**Comment by pratikunterwegs** [comment]
Just coming to this thread after discussion today - happy to help too with packaging up or reviewing, as required.

**Comment by joshwlambert** [comment]
Thanks @CarmenTamayo for posting this discussion, and thanks to @jamesmbaazam for input. Carmen and I have started developing the [{simulist} package as part of the Epiverse-TRACE organisation](https://github.com/epiverse-trace/simulist). 

We will work towards an MVP containing the functionality outlined in the above posts on this discussion. Once it is ready for feedback we will reach out to the rest of the team for input.

Please post comments on the [issues section of the github repo](https://github.com/epiverse-trace/simulist/issues) and not here to keep the work as centralised as possible. I will close this thread once a working version of the package is up and running.

**Comment by joshwlambert** [comment]
The [{simulist} package](https://github.com/epiverse-trace/simulist) is now up and running and approaching an MVP. Therefore, I'm going to close this discussion board and all further comments can be send to the [{simulist} GitHub repository issues page](https://github.com/epiverse-trace/simulist/issues).
